### PR TITLE
🌱 local e2e script

### PR DESCRIPTION
**Pull request by sethp-nr**
This change introduces a `test_e2e_local.sh` script for running e2e
tests locally. Addresses #1435.

Split out from #1546

### CODE REVIEW

**Comment by sethp-nr** [comment]
I was going off the eng notes from that #1435, which include:

> Also, have the kind installed. See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/scripts/setup.sh#L24.

and 

> before running the script is required to export the env var with the k8s version. See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test_e2e.sh#L33 (E.g export KIND_K8S_VERSION=v1.17.0)

Following those instructions, the local script here provided does work for me on macOS. 

> So, we cannot install and set up the kind in this scenario.

 > However, for local dev env people has minikube or any other k8s cluster to work with already.

Would you explain a little more what you mean here? Removing the `kind` dependency for the e2e tests and letting devs bring their own clusters would be a bit more work, I think, and make the environment less similar to the real e2e test.

In my experience it's been acceptable to folks to ask them to install `kind` to do kubernetes related development, and is the growing "default" path for doing these kinds of things. What do you think of keeping the `kind` dependency for now, and if someone wants to use minikube in the future for the local tests they can file a PR for that?

**Comment by sethp-nr** [comment]
I had another thought about the `kind` dependency: would it be better to install a project-local version of the kind binary by way of a `tools` module?

**Comment by camilamacedo86** [comment]
Hi @sethp-nr, 

Sorry, I catcher now. If it works for mac and Linux shows fine. I am testing it now in a mac.

File: scripts/setup.sh [hunk 1]
```diff
@@ -15,13 +15,17 @@
 # limitations under the License.
```
Comment (camilamacedo86) [inline scripts/setup.sh:15]: Could we add a new Makefile target for that? 

.PHONY: test-e2e-local
test-e2e-local: ## It will run the script to install kind and run e2e tests         
	./test_e2e_local.sh

File: test_e2e_local.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+source common.sh
+
+export TRACE=1
+export GO111MODULE=on
+
```
Comment (camilamacedo86) [inline test_e2e_local.sh:24]: Could we export here by default `export KIND_K8S_VERSION=v1.17.0`  and/or in the Makefile?

File: test_e2e_local.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+export KIND_CLUSTER=local-kubebuilder-e2e
+if ! kind get clusters | grep -q $KIND_CLUSTER ; then
+    source "$(pwd)/scripts/setup.sh" ${KIND_K8S_VERSION} $KIND_CLUSTER
+    docker pull gcr.io/kubebuilder/kube-rbac-proxy:v0.5.0
+    kind load --name $KIND_CLUSTER docker-image gcr.io/kubebuilder/kube-rbac-proxy:v0.5.0
+fi
+
+kind export kubeconfig --kubeconfig $tmp_root/kubeconfig --name $KIND_CLUSTER
+export KUBECONFIG=$tmp_root/kubeconfig
+
```
Comment (camilamacedo86) [inline test_e2e_local.sh:40]: We need the trap to ensure that all will be cleaned if an issue is faced.

```shell
# remove running containers on exit
function cleanup() {
    kind delete cluster
}

trap cleanup EXIT
``

File: scripts/setup.sh (hunk 1 above)
Comment (camilamacedo86) [reply scripts/setup.sh:15]: Also could you add new makefile target in : https://github.com/kubernetes-sigs/kubebuilder/blob/master/CONTRIBUTING.md#what-to-do-before-submitting-a-pull-request

File: test_e2e_local.sh (hunk 3 above)
Comment (sethp-nr) [reply test_e2e_local.sh:41]: I found that on my machine, standing up the kind cluster a significant chunk of the total test run time. So for the local script I added in an override that leaves the kind cluster in place between runs: the tests seem very good at either cleaning up after themselves or not conflicting with each other, or both.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -54,14 +54,15 @@ Following the targets that can be used to test your changes locally.
 | make lint |  Check the code implementation | yes   |
 | make test-coverage |  Run coveralls to check the % of code covered by tests | yes   |
 | make check-testdata |  Checks if the testdata dir is updated with the latest changes | yes   |
+| make test-e2e-local |  Runs the CI e2e tests locally | no   |
```
Comment (camilamacedo86) [inline CONTRIBUTING.md:57]: Could you please revert the changes that has no relation with the scope of this pr? (E.g the spaces an just keep what is related to the new target)

File: Makefile [hunk 5]
```diff
@@ -77,11 +77,16 @@ test: ## Run the unit tests (used in the CI)
 
 .PHONY: test-coverage
 test-coverage:  ## Run coveralls
-	# remove all coverage files if exists
+	# remove all coverage files if existss
```
Comment (camilamacedo86) [inline Makefile:80]: revert ^

Comment (sethp-nr) [reply Makefile:80]: oof, good catch!

Can you tell I have one of the awful scissor apple keyboards? :)

**Comment by sethp-nr** [comment]
Thanks @camilamacedo86 for your feedback!
